src/malli/design_draft.clj

```diff
@@ -0,0 +1,96 @@
+(ns malli.design-draft
+  (:import (clojure.lang IPersistentMap)))
+
+
+;;; malli schemas
+
+;; Core schemas are not namespace qualified keywords to make using them user
+;; friendly.
+
+;; Schemas are written using Hiccup style vectors, where order is significant.
+
+;; Every schema definition can have additional options-map after schema name.
+;; Options-map can have arbitrary amount of data, with some keys having special
+;; meaning, like below, which describes a closed map.
+
+[:map {:closed true, :description "My closed map"}]
+
+;; Schema definitions live in registry. Malli provides a core registry for
+;; all the essential schemas. Registry can be extended on the runtime and it's
+;; possible to have as many registries one needs.
+
+;; The actual registry definitions and format is wip, but could be something
+;; like below.
+
+;; Compiles given definition to Schema
+(declare compile)
```

Comment:
I don't see the need for something like this. Instead I would just have `expand` (or `hydrate`?) to turn the hiccup-like data schema into a tree of `Schema` protocol implementors and then those can be compiled to checkers, coercions or generators. I think it is clearer to just create the schema implementations in registry with vanilla `deftype` or `reify`.

 I guess it would be possible to have `Schema` implement checking, coercion and generation directly but any other operations on schemas (e.g. emit SQL table schemas, produce error messages for end users...) would have to be added outside `Schema` anyway. And supporting several operations directly would probably inhibit some optimizations that a separate coercion compiler could do.

Reply:
:+1:

Reply:
Ok, sounds good, `expand` turns data-only schema to tree of `Schema` protocol instances. 

I don't have an opinion yet of should the checking aka validation, coercion and generation be part of the Schema implementation or not. But I think it definitely worth trying out to keep them separate from the Schemas to make multiple usages of Schemas as broad as possible. Lets see how it unfolds.